mrmustard/lab_dev/sampler.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Samplers for measurement devices.
+"""
+
+from __future__ import annotations
+
+from typing import Sequence, Iterable
+
+import numpy as np
+
+from mrmustard import math, settings
+
+from .states import State, Number
+from .circuit_components import CircuitComponent
+from .circuit_components_utils import BtoQ, TraceOut
+
+__all__ = ["Sampler", "PNRSampler", "HomodyneSampler"]
+
+
+class Sampler:
+    r"""
+    A sampler for measurements of quantum circuits.
+
+    Args:
+        meas_outcomes: The measurement outcomes for this sampler.
+        meas_ops: The optional measurement operators of this sampler.
+        prob_dist: An optional probability distribution for this sampler.
+    """
+
+    def __init__(
+        self,
+        meas_outcomes: list[any],
+        meas_ops: CircuitComponent | list[CircuitComponent] | None = None,
+        prob_dist: list[float] | None = None,
+    ):
+        self._meas_ops = meas_ops
+        self._meas_outcomes = meas_outcomes
+        self._prob_dist = prob_dist
+
+    @property
+    def meas_ops(self) -> CircuitComponent | list[CircuitComponent] | None:
+        r"""
+        The measurement operators of this sampler.
+        """
+        return self._meas_ops
+
+    @property
+    def meas_outcomes(self) -> list[any]:
+        r"""
+        The measurement outcomes of this sampler.
+        """
+        return self._meas_outcomes
+
+    @property
+    def prob_dist(self) -> list[float] | None:
+        r"""
+        The probability distribution of this sampler.
+        """
+        return self._prob_dist
+
+    def probabilities(self, state: State | None = None) -> list[float] | None:
+        r"""
+        Returns the probability distribution of this sampler. If ``state`` is provided
+        then will compute the probability distribution w.r.t. the state.
+
+        Args:
+            state: The state to generate the probability distribution with.
+        """
+        self._validate_state(state)
+        if state is not None:
+            dm_state = state.dm()
+            states = [dm_state >> meas_op.dual for meas_op in self.meas_ops]
+            probs = [
+                state.probability if isinstance(state, State) else math.real(state)
+                for state in states
+            ]
+            return probs / sum(probs)
```

Comment:
This will silently re-normalise situations where there is little overlap with the state.  Should a warning or some other message be emitted when sum(probs) isn't close to 1.0?